scenario/node/src/gatewaycontext.ts

```diff
@@ -95,16 +96,17 @@ export class GatewayContext {
         this.closeBlockEvents(listenerName);
         const events = await this.getNetwork().getBlockEvents(options);
         const listener = new CheckpointEventListener<common.Block>(events, async (event: common.Block): Promise<void> => {
-            const blockNumber = event.getHeader()?.getNumber();
-            await this.#checkpointer?.checkpointBlock(BigInt(blockNumber!));
+            const header = assertDefined(event.getHeader(), 'block header');
+            const blockNumber = header.getNumber();
+            await this.getCheckpointer()?.checkpointBlock(BigInt(blockNumber));
```

Comment:
The point of using `getCheckpointer()` is that it guarantees a checkpointer is returned since it throws if one is not present. The reason this isn't obvious is because the implementation and return type of `getCheckpointer()` is wrong. It should be

```typescript
    private getCheckpointer(): Checkpointer {
        return assertDefined(this.#checkpointer, 'checkpointer');
    }
```

There is no need to use the `?` operator when the value can't be undefined.
```suggestion
            await this.getCheckpointer().checkpointBlock(BigInt(blockNumber));
```